projects/packages/waf/src/class-waf-runner.php

```diff
@@ -98,4 +112,82 @@ public static function run() {
 	public static function errorHandler( $code, $message, $file, $line ) { // phpcs:ignore
 		// Intentionally doing nothing for now.
 	}
+
+	/**
+	 * Initializes the WP filesystem.
+	 *
+	 * @return void
+	 * @throws \Exception If filesystem is unavailable.
+	 */
+	protected static function initialize_filesystem() {
+		if ( ! function_exists( '\\WP_Filesystem' ) ) {
+			require_once ABSPATH . 'wp-admin/includes/file.php';
+		}
+
+		if ( ! \WP_Filesystem() ) {
+			throw new \Exception( 'No filesystem available.' );
+		}
+	}
+
+	/**
+	 * Activates the WAF by generating the rules script and setting the version
+	 *
+	 * @return void
+	 */
+	public static function activate() {
+		self::define_mode();
+		if ( ! self::is_allowed_mode( JETPACK_WAF_MODE ) ) {
+			return;
+		}
+		$version = get_option( self::VERSION_OPTION_NAME );
+		if ( ! $version ) {
+			add_option( self::VERSION_OPTION_NAME, self::JETPACK_WAF_VERSION );
+		}
+		self::generate_rules();
+	}
+
+	/**
+	 * Updates the rules.php and the version in site DB
+	 *
+	 * @return void
+	 */
+	public static function update() {
+		self::define_mode();
+		if ( ! self::is_allowed_mode( JETPACK_WAF_MODE ) ) {
+			return;
+		}
+		update_option( self::VERSION_OPTION_NAME, self::JETPACK_WAF_VERSION );
+		self::generate_rules();
+	}
+
+	/**
+	 * Checks if the version has been changed.
+	 *
+	 * @return boolean
+	 */
+	public static function has_version_changed() {
+		$version = get_option( self::VERSION_OPTION_NAME );
+		return self::JETPACK_WAF_VERSION !== $version ? true : false;
+	}
+
+	/**
+	 * Generates the rules.php script
+	 *
+	 * @throws \Exception If filesystem is not available.
+	 * @throws \Exception If file writing fails.
+	 * @return void
+	 */
+	public static function generate_rules() {
+		global $wp_filesystem;
+
+		self::initialize_filesystem();
+
+		// Ensure that the folder exists.
+		if ( ! $wp_filesystem->is_writable( dirname( self::RULES_FILE ) ) ) {
```

Comment:
That makes sense and I will switch it - although again, I pulled the code for handling of filesystem from Jetpack Backups. I may ask them if there was specific reasoning behind it.

Reply:
Come to think of it, I suspect that is_writable may ensure that if the directory does exist, that we have privileges to write to it - but yes, either way the failure will be in the put_contents check. I'll switch it to is_dir because it seems a bit clearer anyhow.